File: leetcode/530.cpp
#include <cstdlib>
#include <iostream>
#include <climits>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
  int min_diff = INT_MAX;
  void dfs(TreeNode *root, int &val) {
    if (root == NULL) {
      return;
    }
    dfs(root->left, val);
    if (val != -1) {
      min_diff = min(min_diff, root->val - val);
    }
    val = root->val;
    dfs(root->right, val);
  }

public:
  int getMinimumDifference(TreeNode *root) {
    min_diff = INT_MAX;
    int prev = -1;
    dfs(root, prev);
    return min_diff;
  }
};

File: leetcode/530.py
# Definition for a binary tree node.
class TreeNode:

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def getMinimumDifference(self, root: TreeNode) -> int:

        def dfs(root: TreeNode):
            nonlocal ans
            nonlocal pre
            if not root:
                return
            dfs(root.left)
            if pre:
                ans = min(ans, abs(root.val - pre.val))
            pre = root
            dfs(root.right)

        ans = float('inf')
        pre = None
        dfs(root)
        return ans